torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1777,6 +1796,14 @@ void ProcessGroupNCCL::watchdogHandler() {
 
       // Clean up completed work
       if (work.isCompleted()) {
+        {
+          // Reset the timeout and first work if the work is completed.
+          std::lock_guard<std::mutex> lock(mtxTimeoutExtension_);
+          if (work.opTimeoutReduce_.count() > 0) {
+            extendedTimeout_ -= work.opTimeoutReduce_;
```

Comment:
Do we really need to keep timeout states in Work objects, opTimeoutReduce_? ProcessGroupNCCL::WorkNCCL::checkTimeout() already allows you to pass an timeout value to replace the default one

Reply:
Instead of using a map to store of Work and timeoutReductionOnComplete_, storing in WorkNCCL is pretty much similar.

You can check the discussion here:
https://github.com/pytorch/pytorch/pull/130905#discussion_r1695416936